### PR TITLE
feat(DTFS2-7659): deal cancellation - gef activity entry - design update

### PR DESCRIPTION
**Pull request by ttbarnes**
## Introduction :pencil2:
This PR updates the "portal activites" design in GEF so that when a deal is cancelled, status tags of "previous" and "new" status appear in the "cancellation" activity.

Note - E2E tests will be in another PR as part of DTFS2-7656 and DTFS2-7656. Currently blocked by another PR.

## Resolution :heavy_check_mark:
- GEF UI:
  - Update the activities GET controller to return a `previousStatus` property/value.
  - Refactor the "activites" mapping and nunjucks template. It is very hard to follow and add additional custom activities with the current structure. The new approach is simpler, easier to read/follow and maintain.

## Miscellaneous :heavy_plus_sign:
- Minor E2E improvements RE activities.

## Screenshot example
Screenshot 2024-12-04 at 15 11 19

### CODE REVIEW

File: dtfs-central-api/src/services/portal/deal.service.add-deal-cancelled-activity.test.ts [hunk 1]
```diff
@@ -61,6 +61,7 @@ describe('PortalDealService - addGefDealCancelledActivity', () => {
           firstName: UKEF.ACRONYM,
         },
         label: PORTAL_ACTIVITY_LABEL.DEAL_CANCELLED,
+        html: PORTAL_ACTIVITY_LABEL.DEAL_CANCELLED,
```
[the remark this answers is not in the document]
Comment (MarRobSoftwire) [reply dtfs-central-api/src/services/portal/deal.service.add-deal-cancelled-activity.test.ts:64]: I've now seen the nunjucks & realised what this is - I'm not sure whats best to do about is as its really quite confusing

File: e2e-tests/tfm/cypress/e2e/journeys/case/activities/activities.spec.js [hunk 2]
```diff
@@ -98,24 +105,29 @@ context('Users can create and submit comments', () => {
...
       activitiesPage.activitiesTimeline().contains('test');
       activitiesPage.activitiesTimeline().contains(userFullName);
-      // ensures that is filtered out
+
+      // ensures the activity is filtered out
       activitiesPage.activitiesTimeline().contains('Not a comment').should('not.exist');
     });
 
     it('should not be allowed to add comment over 1000 characters', () => {
       const longComment = 'aaaaaaaaaa'.repeat(101);
```
[the remark this answers is not in the document]
Comment (ttbarnes) [reply e2e-tests/tfm/cypress/e2e/journeys/case/activities/activities.spec.js:125]: Great spot, didn't see. Updated

File: gef-ui/templates/partials/application-activity.njk [hunk 3]
```diff
@@ -21,54 +22,86 @@
...
+        {
+          label: activity.label,
+          text: activity.text,
+          datetime: activity.datetime,
+          byline: '',
+          html: facilityHtml
+        }
+      %}
+
+      {% elif activity.html === 'Deal cancelled' %}
```
[the remark this answers is not in the document]
Comment (ttbarnes) [reply gef-ui/templates/partials/application-activity.njk:73]: I completely agree RE the `html` property - very hesitant to change this at this moment in time - same with other activity properties. We can rename them, but we'd need either data migration, or messy conditional legacy support in nunjucks. I think for the purpose of deal cancellation, we should stick to the existing structure/naming.

I don't find the nunjucks file hard to follow but i'll look into mapping server side which would also be my preference.

File: dtfs-central-api/src/services/portal/deal.service.add-deal-cancelled-activity.test.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ttbarnes) [reply dtfs-central-api/src/services/portal/deal.service.add-deal-cancelled-activity.test.ts:64]: Completely agree, it is confusing - am looking into improving the consumption of this on the UI side at least.

Ideally we would rename `html` to something like `componentName` / `componentType`, but I don't think we should do this as part of this PR/deal cancellation.

File: gef-ui/templates/partials/application-activity.njk (hunk 3 above)
[the remark this answers is not in the document]
Comment (ttbarnes) [reply gef-ui/templates/partials/application-activity.njk]: I've completely refactored the mapping and the nunjucks side of it. No use of `html` now and various other readability/maintainability improvements 👍 

Probably easiest to resolve this comment and start any comments in a new thread

File: gef-ui/server/controllers/activities-controller/index.test.js [hunk 4]
```diff
@@ -1,19 +1,25 @@
...
+import { fromUnixTime, format } from 'date-fns';
+import { timeZoneConfig, DATE_FORMATS } from '@ukef/dtfs2-common';
 import { mapPortalActivities, getPortalActivities } from '.';
 import api from '../../services/api';
 import mocks from '../mocks';
 
 jest.mock('../../services/api');
 
+const timestamp = 1733311320;
+const date = fromUnixTime(new Date(timestamp));
```
Comment (MarRobSoftwire) [inline gef-ui/server/controllers/activities-controller/index.test.js:10]: This isn't doing what it seems...
-  `fromUnixTime` should take a timestamp not a date, its only functioning because Dates are coerced into numbers automatically
- Moreover `new Date(1733311320)` returns januray 1970, not December 2024 as you might expect

```suggestion
const timestamp = 1733311320;
const date = fromUnixTime(timestamp);
```

File: gef-ui/templates/partials/application-activity.njk [hunk 5]
```diff
@@ -21,54 +22,47 @@
...
   <h1 class="govuk-heading-xl" data-cy="main-activity-heading">Activity and comments</h1>
 
-    {% set portalActivitiesWithHTML = [] %}
+  {# 
+    NOTE: we need to use MOJ timeline HTML rather than the nunjucks component.
+    This is because we have conditional custom HTML for different timeline items/activities
+    See the "activity-html" / activityHtml component for examples.
+    The alternative to this is to have complex array mapping, object creation and creation.
+    inside this nunjucks template, which we found to be very hard to read and maintain.
+  #}
```
Comment (MarRobSoftwire) [inline gef-ui/templates/partials/application-activity.njk:31]: I think it might be possible to use the nunjucks component & abstract the activity html away
e.g. 

```
  {% set portalActivitiesWithHTML = [] %}

  {% for activity in portalActivities %}

    {% set activityToAdd = {
        label: activity.label,
        text: activity.text,
        datetime: activity.datetime,
        byline: activity.byline,
        html: activityHtml.render(activity)
      }
    %}

    {% set portalActivitiesWithHTML = (portalActivitiesWithHTML.push(activityToAdd), portalActivitiesWithHTML) %}
  {% endfor %}

  {{ mojTimeline({
      items: portalActivitiesWithHTML,
      attributes: {
        'data-cy': 'portal-activities-timeline'
      }
    }) }}
```

Comment (ttbarnes) [reply gef-ui/templates/partials/application-activity.njk:31]: Pretty sure I tried this or similar and it wouldn't work because of the MOJ component and `html: activityHtml.render(activity)`. I.e it would not actually render the HTML.

Will double check.

File: gef-ui/server/controllers/activities-controller/index.test.js [hunk 6]
```diff
@@ -59,11 +61,10 @@ describe('mapPortalActivities', () => {
...
       {
-        label: { text: 'Automatic inclusion notice submitted to UKEF' },
-        text: '',
-        datetime: { timestamp: fromUnixTime(1638458265), type: 'datetime' },
-        byline: { text: 'Bob Smith' },
-        html: '',
+        title: 'Automatic inclusion notice submitted to UKEF',
+        date: format(date, DATE_FORMATS.D_MMMM_YYYY),
+        time: format(date, DATE_FORMATS.H_MMAAA),
+        byline: 'Bob Smith',
```
[the remark this answers is not in the document]
Comment (ttbarnes) [reply gef-ui/server/controllers/activities-controller/index.test.js:67]: Good shout - didn't want to get _too_ deep into cleaning up - will go through all of the assertions to be DRY.

File: e2e-tests/gef/cypress/e2e/facility/review-ukef-approval/facility-end-date-disabled/review-ukef-approval-with-issued-and-unissued-facilities-MIA-and-resubmit-to-ukef.spec.js [hunk 7]
```diff
@@ -501,8 +501,8 @@ context('Check activity feed', () => {
       applicationActivities
         .facilityActivityCheckedBy(unissuedFacilitiesArray[0].ukefFacilityId)
         .contains(`Checked by ${BANK1_CHECKER1.firstname} ${BANK1_CHECKER1.surname}`);
-      applicationActivities.facilityActivityUnissuedTag(unissuedFacilitiesArray[0].ukefFacilityId).contains('Unissued');
-      applicationActivities.facilityActivityIssuedTag(unissuedFacilitiesArray[0].ukefFacilityId).contains('Issued');
+      applicationActivities.previousStatusTag(unissuedFacilitiesArray[0].ukefFacilityId).contains('Unissued');
+      applicationActivities.newStatusTag(unissuedFacilitiesArray[0].ukefFacilityId).contains('Issued');
```
Comment (Zainzzkk) [inline e2e-tests/gef/cypress/e2e/facility/review-ukef-approval/facility-end-date-disabled/review-ukef-approval-with-issued-and-unissued-facilities-MIA-and-resubmit-to-ukef.spec.js:505]: could this use assertText?

File: gef-ui/templates/partials/application-activity.njk (hunk 5 above)
Comment (ttbarnes) [reply gef-ui/templates/partials/application-activity.njk:31]: Had a think about this @MarRobSoftwire - even if the suggestion above works, I don't think we should do it.

Really not a fan of constructing and pushing objects in a nunjucks file, same with arrays. I think it's much cleaner to have it simply consuming data from the controller.

File: dtfs-central-api/src/services/portal/deal.service.ts [hunk 8]
```diff
@@ -69,6 +69,7 @@ export class PortalDealService {
           firstName: UKEF.ACRONYM,
         },
         label: PORTAL_ACTIVITY_LABEL.DEAL_CANCELLED,
+        html: PORTAL_ACTIVITY_LABEL.DEAL_CANCELLED,
```
[the remark this answers is not in the document]
Comment (ttbarnes) [reply dtfs-central-api/src/services/portal/deal.service.ts:72]: This has been removed

File: gef-ui/component-tests/partials/application-activity.component-test.js [hunk 9]
```diff
@@ -40,7 +36,19 @@ describe(page, () => {
...
+    wrapper.expectText(`[data-cy="activity-${mockActivity.title}-title"]`).toRead(mockActivity.title);
+  });
+
+  it('should render an activity byline', () => {
+    wrapper.expectText(`[data-cy="activity-${mockActivity.title}-byline"]`).toRead(`by ${mockActivity.byline}`);
+  });
+
+  it('should render an activity date', () => {
+    wrapper.expectText(`[data-cy="activity-${mockActivity.title}-time"]`).toRead(`${mockActivity.date} ${mockActivity.time}`);
+  });
```
Comment (MarRobSoftwire) [inline gef-ui/component-tests/partials/application-activity.component-test.js:53]: can you check this renders the different activity types correctly

Comment (MarRobSoftwire) [reply gef-ui/component-tests/partials/application-activity.component-test.js:53]: e.g. check for `datacy="previous-status-tag"`

File: gef-ui/templates/partials/application-activity.njk (hunk 5 above)
Comment (MarRobSoftwire) [reply gef-ui/templates/partials/application-activity.njk:31]: Happy to leave this then, & see what you mean, though I think this does bring with it a burden of maintaining the component if the timeline is ever updated

File: gef-ui/component-tests/partials/application-activity.component-test.js (hunk 9 above)
Comment (ttbarnes) [reply gef-ui/component-tests/partials/application-activity.component-test.js:53]: Have manually checked activities AFAIK - all seems to be ok. Fixed an issue picked up in E2E tests. Screenshot of the latest:

Screenshot 2024-12-10 at 15 06 18

File: gef-ui/templates/partials/application-activity.njk (hunk 5 above)
Comment (ttbarnes) [reply gef-ui/templates/partials/application-activity.njk:31]: I do agree, I don't believe there has been an update to the timeline component - if there is, we'll at least only have to manually update it in one place.

I think there's more pros than cons to _not_ using the component here.

File: gef-ui/component-tests/partials/application-activity.component-test.js (hunk 9 above)
Comment (Zainzzkk) [reply gef-ui/component-tests/partials/application-activity.component-test.js:53]: could you check if the previous status and the new status tag exist in this test?

File: gef-ui/server/controllers/activities-controller/index.test.js [hunk 10]
```diff
@@ -1,21 +1,27 @@
...
+const timestamp = 1733311320;
+const date = fromUnixTime(timestamp);
+
+const mockAuthor = {
+  firstName: 'Bob',
+  lastName: 'Smith',
+  _id: 12345,
+};
+
 const dealSubmissionActivity = [
```
Comment (Zainzzkk) [inline gef-ui/server/controllers/activities-controller/index.test.js:18]: could this go into mocks?

File: gef-ui/templates/_macros/application-activity/activity-html.njk [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+  {% set ukefDealId = params.ukefDealId %}
+  {% set previousStatus = params.previousStatus %}
+  {% set facilityType = params.facilityType %}
+  {% set facilityId = params.facilityId %}
+  {% set ukefFacilityId = params.ukefFacilityId %}
+  {% set maker = params.maker %}
+  {% set checker = params.checker %}
+
+  {% if title === 'Bank facility stage changed' %}
+
```
Comment (MarRobSoftwire) [inline gef-ui/templates/_macros/application-activity/activity-html.njk:18]: Does this value come from the database? If so is this the value `title` takes on existing deals?

File: gef-ui/component-tests/partials/application-activity.component-test.js (hunk 9 above)
Comment (ttbarnes) [reply gef-ui/component-tests/partials/application-activity.component-test.js:53]: Sorry chaps, forgot to update component tests. Have updated.

File: gef-ui/templates/_macros/application-activity/activity-html.njk (hunk 11 above)
Comment (ttbarnes) [reply gef-ui/templates/_macros/application-activity/activity-html.njk:18]: It's saved as `label` in the DB as per other activities. The MOJ component has a `title` element so in the UI's controller we map `title` to `label`.

**Review by Zainzzkk — APPROVED** [review]

File: gef-ui/component-tests/partials/application-activity.component-test.js (hunk 9 above)
Comment (ttbarnes) [reply gef-ui/component-tests/partials/application-activity.component-test.js]: Marking as resolved for merge

File: gef-ui/templates/_macros/application-activity/activity-html.njk (hunk 11 above)
Comment (ttbarnes) [reply gef-ui/templates/_macros/application-activity/activity-html.njk:18]: Marking as resolved for merge and to help clear the backlog.

Happy to change this post-merge if required @MarRobSoftwire
